[D2] 기프티콘 등록 화면 디테일 구현

AhnSangHee opened the pull request:
# 개요
- 🔗  이슈링크 : resolve #62 


# 변경사항
## 💻 작업 내용
- 커스텀 이미지 피커 → PHPicker 사용하도록 변경
- 스크롤뷰 내부에 있는 버튼을 밖으로 뺌
- 선택된 사진 가져와서 기프티콘 등록 화면에 표시
- 기프티콘 등록 화면에서 원본 보기 버튼 클릭하면 원본 볼 수 있도록 구현
- 상단 내비게이션 바에 닫기 버튼, 내비게이션 타이틀 추가

## 📱 미리보기
원래는 사진을 선택하면 기프티콘을 분석하는 팝업창이 임시로 떠야하는데, 지금은 임시로 알림창이 뜨도록 구현했습니다.
작업 전|작업 후
---|---
![Simulator Screen Recording - iPhone 13 Pro Max - 2022-07-21 at 16 17 07](https://user-images.githubusercontent.com/95578975/180153296-18ecfddb-a9ca-4740-8696-70e8e9bcf57c.gif)|![Simulator Screen Recording - iPhone 13 Pro Max - 2022-07-21 at 16 14 49](https://user-images.githubusercontent.com/95578975/180153070-6355f6df-1d90-4281-99c7-c95f8d9d8933.gif)

AhnSangHee committed:
:bulb: 스크롤뷰 바깥에 버튼 배치 (#62)

AhnSangHee committed:
:bulb: 기프티콘 이미지 원본 보기 버튼 델리게이트 추가 (#62)

AhnSangHee committed:
:bulb: 기프티콘 사진 선택 화면 PHPicker 사용하도록 변경 (#62)

AhnSangHee committed:
:bulb: 갤러리에서 선택한 사진을 기프티콘 등록 화면 이미지뷰에 표시 (#62)

AhnSangHee committed:
:recycle: 사용하지 않는 파일 제거 (#62)

AhnSangHee committed:
:sparkles::bulb: 원본 보기 버튼 누르면 원본 사진 뜨도록 구현 (#62)

szzang-kr committed:
:recycle: NavigationBar 수정

-- 타이틀을 생성자로 받은 경우 바로 셋팅하도록 변경
-- NavigationBar가 자체의 height을 가지도록 변경

AhnSangHee committed:
:bulb: 기프티콘 원본 보기 뷰 background 색상 변경 (#62)

szzang-kr committed:
:recycle: 네비게이션바 왼쪽 버튼 아이템 수정

AhnSangHee committed:
:bulb: 기프티콘 등록 화면 내비게이션 바 구현 (#62)

- 닫기 버튼, 내비게이션 바 타이틀 추가

AhnSangHee committed:
:recycle: 접근 제어자 수정 (#62)

AhnSangHee committed:
:recycle: 접근 제어자 수정 (#62)

szzang-kr commented on Projects/App/Sources/Main/MainViewController.swift:
```diff
@@ -139,4 +121,36 @@ final class MainViewController: BaseViewController<MainViewModelProtocol> {
+                }
+            }
+        }
+        dismiss(animated: true)
```
💭 비즈니스 로직은 ViewModel에 둘 수 있을 것 같은데요, 어떻게 두면 좋을지 함께 이야기 해보아용

szzang-kr commented on Projects/App/Sources/Register/GiftionImageViewController.swift:
```diff
@@ -30,6 +30,8 @@ final class GiftionImageViewController: UIViewController {
     }
     
     private func configure() {
+        view.backgroundColor = .white
```
```suggestion
        view.backgroundColor = .ddip(.neutralWhite))
```

designSystem을 사용해주세요 😅

AhnSangHee replied:
흰색의 경우 모두 designSystem을 사용하도록 변경했습니다~!

szzang-kr commented on Projects/App/Sources/Register/RegisterGifticonViewController.swift:
```diff
@@ -8,34 +8,56 @@
+        
+        navigationBar.leftButtonTapEvent.subscribe(onNext: {
+            self.dismiss(animated: true)
+        }).disposed(by: disposeBag)
```
onNext의 내용은 closure이기 때문에 strong reference를 사용하게되면 reference cycle을 발생시킬 수 있습니다. 😅

AhnSangHee replied:
코멘트 달아주신 부분 수정하여 반영했습니다!!!

szzang-kr replied:
@ios-h

[내비게이션 버튼 탭 이벤트에 reference cycle 제거 ](https://github.com/mash-up-kr/GGiriGGiri_iOS/pull/64/commits/457376476394cc5b7c915aec131439b7da436694)
[delegate를 weak var로 변경 ](https://github.com/mash-up-kr/GGiriGGiri_iOS/pull/64/commits/0d8101a3f69cdd090b92bbed07a8e7d459a0dec0) 두 커밋에서 확인 완료했습니다 😆

szzang-kr reviewed (approved):
수고하셨습니다 !! 😆

AhnSangHee committed:
:recycle: 디자인 시스템 컬러 사용하도록 변경 (#62)

kimkyunghun3 commented on Projects/App/Sources/Main/MainViewController.swift:
```diff
@@ -124,12 +111,7 @@ final class MainViewController: BaseViewController<MainViewModelProtocol> {
-                navigationController.modalPresentationStyle = .fullScreen
-                self.present(navigationController, animated: true)
-            }
+            self.presentPhotoPicker()
```
self 사용에 대해서는 통일성을 유지하면 좋을거같습니다!

kimkyunghun3 commented on Projects/App/Sources/Register/View/RegisterGifticonView.swift:
```diff
@@ -69,10 +58,21 @@ final class RegisterGifticonView: BaseView {
         registerButton.snp.makeConstraints {
-            $0.top.equalTo(registerGifticonDDipInfoView.snp.bottom).offset(48)
-            $0.leading.trailing.equalTo(infoMessageView)
+            $0.top.equalTo(scrollView.snp.bottom).offset(46)
```
요렇게 숫자로 대입하면 여러 기기에서도 대응이 가능한가요?

AhnSangHee replied:
넵! 스크롤뷰는 스크롤뷰만의 높이를 갖고, 높이가 정해진 스크롤뷰로부터의 offset 만큼 상수를 주는 것이기 때문에 기기와 상관 없이 대응 가능합니다!😀

kimkyunghun3 commented on Projects/App/Sources/Main/MainViewController.swift:
```diff
@@ -139,4 +121,37 @@ final class MainViewController: BaseViewController<MainViewModelProtocol> {
+extension MainViewController: PHPickerViewControllerDelegate {
+    func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
+
+        if let itemProvider = results.first?.itemProvider, itemProvider.canLoadObject(ofClass: UIImage.self) {
```
itemProvider가 무엇을 말하는걸까요? 이름을 보았을 떄 정확히 어떤 역할인지 파악이 어렵네요!

AhnSangHee replied:
results.first?.itemProvider의 itemProvider는 인스턴스 프로퍼티로 [여기](https://developer.apple.com/documentation/photokit/phpickerresult/3606600-itemprovider)를 보시면 자세하게 보실 수 있습니다! 간략하게 말씀해드리자면 선택된 asset을 representation하는 역할을 합니다. 애플의 공식 예제 프로젝트도 itemProvider로 네이밍을 했기에 저도 itemProvider라고 네이밍했습니다!

kimkyunghun3 reviewed (changes requested):
고생했습니다~~

AhnSangHee committed:
:recycle: 내비게이션 버튼 탭 이벤트에 reference cycle 제거 (#62)

AhnSangHee committed:
:recycle: delegate를 weak var로 변경 (#62)

- delegate가 var로 선언되어 strong reference로 인해 RegisterGifticonVC에서 deinit이 호출되지 않는 문제 해결

AhnSangHee committed:
:recycle: 코드 컨벤션 수정 (#62)
